### PR TITLE
Feature/removed dynamic values

### PR DESCRIPTION
**Pull request by zenangst**

### CODE REVIEW

File: Source/FORMDataSource.h [hunk 1]
```diff
@@ -65,6 +65,7 @@ typedef UICollectionViewCell * (^FORMFieldConfigureCellForItemAtIndexPath)(FORMF
 - (NSInteger)numberOfFields;
 - (NSArray *)forms;
 - (NSDictionary *)valuesDictionary;
+- (NSDictionary *)removedDynamicValues;
```
Comment (NSElvis) [inline Source/FORMDataSource.h]: Maybe we should make this one `removedValues` too

File: Source/FORMDataSource.m [hunk 2]
```diff
@@ -513,36 +513,60 @@ - (void)fieldCell:(UICollectionViewCell *)fieldCell updatedWithField:(FORMField
...
-                    if ([currentSection.position integerValue] > [section.position integerValue]) {
-                        currentSection.position = @([currentSection.position integerValue] - 1);
+                [self updateSectionPosition:section];
+
+                NSMutableArray *removedKeys = [NSMutableArray new];
+                __block NSDictionary *parsed;
+                [self.valuesDictionary enumerateKeysAndObjectsUsingBlock:^(NSString *key, id obj, BOOL *stop) {
+                    if ([key hasPrefix:section.sectionID]) {
+                        parsed = [key hyp_parseRelationship];
+                        NSString *newKey = [NSString stringWithFormat:@"%@[%@].%@", [parsed objectForKey:@"relationship"], @(self.formsManager.removedValues.count), [parsed objectForKey:@"attribute"]];
```
Comment (NSElvis) [inline Source/FORMDataSource.m]: https://github.com/hyperoslo/NSString-HYPRelationshipParser/issues/7 :)

File: Source/FORMDataSource.m [hunk 3]
```diff
@@ -513,36 +513,60 @@ - (void)fieldCell:(UICollectionViewCell *)fieldCell updatedWithField:(FORMField
...
+                        [self.formsManager.values removeObjectForKey:key];
+                        [removedKeys addObject:newKey];
                     }
-                }
+                }];
+
+                parsed = [[section.sectionID stringByAppendingString:@".name"] hyp_parseRelationship];
```
Comment (NSElvis) [inline Source/FORMDataSource.m]: Appending `.name`?

File: Source/FORMDataSource.m [hunk 4]
```diff
@@ -513,36 +513,60 @@ - (void)fieldCell:(UICollectionViewCell *)fieldCell updatedWithField:(FORMField
...
                 }
+
+                NSDictionary *updatedValueKeys = [self updateValueKeys:[self.valuesDictionary allKeys]];
+                NSDictionary *currentValues = self.valuesDictionary;
+
+                [self removeDynamicKeysForSection:section];
+
+                [currentValues enumerateKeysAndObjectsUsingBlock:^(NSString *key, id obj, BOOL *stop) {
+                    if (updatedValueKeys[key]) {
+                        [self.formsManager.values setObject:obj forKey:updatedValueKeys[key]];
```
Comment (NSElvis) [inline Source/FORMDataSource.m]: I guess 
I think we could use the other notation here

```objc
self.formsManager.values[updatedValueKeys[key]] = obj;
```

File: Source/FORMDataSource.m (hunk 2 above)
Comment (zenangst) [inline Source/FORMDataSource.m]: :100:

File: Source/FORMDataSource.m (hunk 3 above)
Comment (zenangst) [inline Source/FORMDataSource.m]: I just used `.name`, no real reference to anything. But `hyp_parseRelationship` wouldn't parse it without the suffix.

File: Source/FORMDataSource.m (hunk 4 above)
Comment (zenangst) [inline Source/FORMDataSource.m]: Fixed - https://github.com/hyperoslo/Form/commit/6e6d19a003cd28026237d24bb1ed0cfabf2d4307

File: Source/FORMDataSource.h (hunk 1 above)
Comment (zenangst) [inline Source/FORMDataSource.h]: I thought about that but it's only the dynamic values that gets added to that property to it made sense to make it more explicit. From an API perspective it makes sense but `removedValues` is shorter and reads nice.

Comment (NSElvis) [inline Source/FORMDataSource.h]: Yes please, also in the future adding more functionality to removedValues while been backwards compatible should be easier to do without breaking the API.

Comment (zenangst) [inline Source/FORMDataSource.h]: But for consistency, wouldn't `removedValuesDictionary` be a better name?

Comment (NSElvis) [inline Source/FORMDataSource.h]: I've been thinking that we should go for `values` and `removedValues`, like you don't say

"indexInteger"

Comment (zenangst) [inline Source/FORMDataSource.h]: Fixed :)

**Comment by zenangst** [comment]
:star:

**Pull request merged**